"Testing environments" page translation (#396)
reactjs/ru.reactjs.org: content/docs/testing-environments.md

---
id: testing-environments
title: Среды тестирования
permalink: docs/testing-environments.html
prev: testing-recipes.html
---

> <!-- Этот документ предназначен для людей, которые знают, как использовать JavaScript и умеют писать тесты. Он показывает различия в средах тестирования React-компонентов и как эти различия повлияют на тесты, которые они пишут. Документ, в основном, рассказывает про основанные на web react-dom-компоненты и кратко упоминает про другие рендереры. -->

Reviewer: Я также бы предложил подумать о вариантах перевода последнего предложения, как-то утекает смысл, на мой взгляд. Хотел бы предложить свой вариант: `Документ в больше части описывает react-dom-компоненты в вебе, но и кратко упоминает про другие рендереры.`, но он тоже какой-то черствый получился.
```suggestion
<!-- Этот документ предназначен для людей, которые знают, как использовать JavaScript и умеют писать тесты. Документ рассказывает о различиях в средах тестирования React-компонентов и как эти различия влияют на тесты, которые они пишут. Документ, в основном, рассказывает про основанные на web react-dom-компоненты и кратко упоминает про другие рендереры. -->
```

Author: Документ, по большей части, про react-dom-компоненты для веб(-платформы) и частично про другие рендереры.

- Jest хорошо совместим с React-проектами, поддерживает такие возможности, как фиктивные [модули](#mocking-modules) и [таймеры](#mocking-timers), работа с [`jsdom`](#mocking-a-rendering-surface). **Если вы используете Create React App, [Jest уже предустановлен](https://facebook.github.io/create-react-app/docs/running-tests) с полезными настройками по-умолчанию.**
- Библиотеки, такие как [mocha](https://mochajs.org/#running-mocha-in-the-browser) хорошо работают в среде с настоящим браузером и помогут в тестах, которым она требуется.
- «Сквозные» тесты нужны для тестирования длинных последовательностей действий через несколько страниц и требуют [другой настройки](#end-to-end-tests-aka-e2e-tests).

### Подмена области рендеринга {#mocking-a-rendering-surface}

Тесты часто запускаются в среде, у которой нет доступка к реальной области рендеринга, например браузера. В этом случае мы рекомендуем имитировать работу браузера с помощью [`jsdom`](https://github.com/jsdom/jsdom) – легковесной реализации браузера, которая запускается в Node.js.

> В большинстве слуаев, `jsdom` ведёт себя как обычный браузер, но без некоторых возможностей, например [разметки и навигации](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform). Несмотря на это, `jsdom` полезен в большинстве тестов web-компонентов, так как выполняется быстрее, чем запуск браузера для каждого отдельного теста и запускается рядом с вашими тестами, что позволяет писать код проверки предполагаемого вывода на отрендеренную DOM.

Reviewer: Мне кажется, что перевод `web-based components` как `web-компоненты` может запутать, так как есть отдельная тема о веб-компонентах. Здесь речь о компонентах, используемые в вебе, но я не смог подобрать подходящие варианты перевода.

```suggestion
В большинстве слуаев, `jsdom` ведёт себя как обычный браузер, но без некоторых возможностей, например [разметки и навигации](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform). Несмотря на это, `jsdom` полезен в большинстве тестов компонентов в web, так как выполняется быстрее, чем запуск браузера для каждого отдельного теста, и запускается рядом с вашими тестами, что позволяет писать код для проверки и утвеждрения отрендеренного DOM.
```

Author: - Предлагаю: ...**полезен в большинстве тестов компонентов для веб(-платформы)**...
- DOM – это модель, соответственно женский род, но я не знаю, возможно, по-русски принято использовать термин DOM в мужском?
- "проверка и утверждение" для стороннего читателя бессмыслица получается. Условно, мы сравниваем то, как отрендерился компонент в DOM, с тем, что (мы утверждаем/предполагаем/assert) должно быть.
Предлагаю: **что позволяет писать код для сравнения отрендеренной(-го) DOM с ожидаемым результатом.**

Полностью:
Несмотря на это, `jsdom` **полезен в большинстве тестов компонентов для веб(-платформы)**, так как выполняется быстрее, чем запуск браузера для каждого отдельного теста, и запускается рядом с вашими тестами, **что позволяет писать код для сравнения отрендеренной(-го) DOM с ожидаемым результатом.**

Reviewer 2: > DOM – это модель, соответственно женский род, но я не знаю, возможно, по-русски принято использовать термин DOM в мужском?

Вроде логично, но в реальности не видел, чтобы эта аббревиатура была женского рода. Тут не все так однозначно, как со многими заимствованиями. Мне кажется, это пошло от того, что DOM - интерфейс (API), а не "непонятная модель", то есть Document Object Model - название этого API, поэтому в русскоязычных текстах пишут ещё "Модель DOM" или "Дерево DOM", чтобы было понятнее, что имеется в виду. В общем, предлагаю, оставить мужской род здесь, потому что везде только видно его.

>  и запускается рядом с вашими тестами,

Он также запускается...

Т.е. нужно отделить, иначе предложение слишком большое.

Author: Несмотря на это, `jsdom` полезен в большинстве тестов компонентов для веб, так как выполняется быстрее, чем запуск браузера для каждого отдельного теста. Он также запускается рядом с вашими тестами, что позволяет писать код для сравнения отрендеренного DOM с ожидаемым результатом.

Как настоящий браузер, `jsdom` имитируют действия пользователя: тесты могут создавать события на DOM-узлах, наблюдать за ними и проверять предпологаемые побочные эффекты от этих действий [<small>(пример)</small>](/docs/testing-recipes.html#events).

> Большое количество UI-тестов можно написать описанным выше способом: Jest запускает тесты, компонент рендерится в `jsdom`, действия пользователя описываются в виде последовательности браузерных событий и оборачивается вспомогательной функцией `act()` [<small>(пример)</small>](/docs/testing-recipes.html). Кстати, много тестов для библиотеки React написано таким образом.

Reviewer: ```suggestion
Большое количество UI-тестов можно написать описанным выше способом: Jest запускает тесты, компонент рендерится в `jsdom`, действия пользователя описываются в виде последовательности браузерных событий и оборачиваются вспомогательной функцией `act()` [<small>(пример)</small>](/docs/testing-recipes.html). Кстати, много тестов для библиотеки React написано таким образом.
```

> Если вы пишете библиотеку, которая в основном тестирует поведение специфичное для браузера, например, работа с разметкой страницы или настоящие элементы `input`, используйте фреймворк, такой как, [mocha.](https://mochajs.org/)

Reviewer: ```suggestion
Если вы пишете библиотеку, которая в основном тестирует поведение специфичное для браузера, например, работа с разметкой страницы или настоящие элементы `input`, используйте фреймворк, такой как [mocha.](https://mochajs.org/)
```

> В среде, где вы _не можете_ имитировать DOM (например, тестирование React Native-компонентов в Node.js), используйте [вспомогательные функции имитации события](https://reactjs.org/docs/test-utils.html#simulate), чтобы имитировать взаимодействие с элементами. В качестве альтернативы, можно использовать вспомогательную функцию `fireEvent` из [`@testing-library/react-native`](https://testing-library.com/docs/native-testing-library).

Reviewer: ```suggestion
В среде, где вы _не можете_ имитировать DOM (например, тестирование React Native-компонентов в Node.js), используйте [вспомогательные функции имитации события](https://reactjs.org/docs/test-utils.html#simulate), чтобы имитировать взаимодействие с элементами. В качестве альтернативы можно использовать вспомогательную функцию `fireEvent` из [`@testing-library/react-native`](https://testing-library.com/docs/native-testing-library).
```

Фреймворки, такие как [Cypress](https://www.cypress.io/), [puppeteer](https://github.com/GoogleChrome/puppeteer) и [webdriver](https://www.seleniumhq.org/projects/webdriver/) полезны для запуска [«сквозных» тестов](#end-to-end-tests-aka-e2e-tests).

### Подмена функций {#mocking-functions}

> При написании тестов, иногда требуется подменить части кода, которые не имеют аналога внутри среды тестирования (например, проверка статуса через `navigator.onLine` внутри Node.js). Тесты также могут «следить» за некоторыми функциями и видеть, как другие части теста взаимодействуют с ними. В этом случае полезно иметь возможность выборочно подменить эти функции на удобные для тестирования версии.

Reviewer: ```suggestion
При написании тестов иногда требуется подменить части кода, которые не имеют аналога внутри среды тестирования (например, проверка статуса через `navigator.onLine` внутри Node.js). Тесты также могут «следить» за некоторыми функциями и видеть, как другие части теста взаимодействуют с ними. В этом случае полезно иметь возможность выборочно подменять эти функции на удобные для тестирования версии.
```

Это особенно удобно при получении данных. Обычно, предпочтительнее использовать «подставные» данные для тестов, чтобы избежать замедления и странного поведения, чем получать данные из настоящих API. [<small>(пример)</small>](/docs/testing-recipes.html#data-fetching). Это сделает тесты более предсказуемыми. Библиотеки, такие как, [Jest](https://jestjs.io/) и [sinon](https://sinonjs.org/), поддерживают подмену функций. В «сквозных» тестах, подменить сетевой интерфейс будет тяжелее, но вы, возможно, захотите протестировать и настоящие API в таких тестах.

### Фиктивные модули {#mocking-modules}

Некоторые компоненты имеют зависимости в виде модулей, могут неправильно работать внутри тестовой среды или совсем не нужны при тестировании. Выборочная подмена таких модулей на фиктивные может быть полезна [<small>(пример)</small>](/docs/testing-recipes.html#mocking-modules).

В Node.js исполнители тестов, например Jest [поддерживает фиктивные модули](https://jestjs.io/docs/ru/manual-mocks). Вы можете также использовать библиотеки, такие как [`mock-require`](https://www.npmjs.com/package/mock-require).

### Фиктивные таймеры {#mocking-timers}

> Компоненты могут использовать функции, связанные с временем, такие как `setTimeout`, `setInterval` или `Date.now`. В средах тестирования подмена таких функции на фиктивные, позволит «двигаться» во времени и может быть полезной. Это отличный способ ускорить выполнение тестов! Тесты с таймерами будут по-прежнему выполнятся по порядку, только быстрее. [<small>(пример)</small>](/docs/testing-recipes.html#timers). Большинство фреймворков, включая [Jest](https://jestjs.io/docs/en/timer-mocks), [sinon](https://sinonjs.org/releases/v7.3.2/fake-timers/) и [lolex](https://github.com/sinonjs/lolex), позволяют подменять таймеры в тестах.

Reviewer: ```suggestion
Компоненты могут использовать функции, связанные с временем, такие как `setTimeout`, `setInterval` или `Date.now`. В средах тестирования может быть полезно подменить такие функции на фиктивные, что позволит «двигаться» во времени. Это отличный способ ускорить выполнение тестов! Тесты с таймерами будут по-прежнему выполнятся по порядку, только быстрее. [<small>(пример)</small>](/docs/testing-recipes.html#timers). Большинство фреймворков, включая [Jest](https://jestjs.io/docs/en/timer-mocks), [sinon](https://sinonjs.org/releases/v7.3.2/fake-timers/) и [lolex](https://github.com/sinonjs/lolex), позволяют подменять таймеры в тестах.
```

> Иногда фиктивные таймеры не нужны. Например, при тестировании анимации или взаимодействии с удаленным API, который чувствителен к времени (имеет ограничение по частоте доступа). Библиотеки с фиктивными таймерами позволяют включать и выключать их для одного теста или набора в целом, так что вы можете явно указать, как эти тесты будут исполняться.

Reviewer: ```suggestion
Иногда фиктивные таймеры не нужны. Например, при тестировании анимации или взаимодействии с удалённым API, который чувствителен к времени (имеет ограничение по частоте доступа). Библиотеки с фиктивными таймерами позволяют включать и выключать их для одного теста или набора в целом, так что вы можете явно указать как эти тесты будут исполняться.
```
